src/hotspot/share/memory/archiveBuilder.cpp

```diff
@@ -159,37 +161,40 @@ void ArchiveBuilder::SourceObjList::relocate(int i, ArchiveBuilder* builder) {
   _ptrmap.iterate(&relocator, start, end);
 }
 
-ArchiveBuilder::ArchiveBuilder(DumpRegion* mc_region, DumpRegion* rw_region, DumpRegion* ro_region)
-  : _rw_src_objs(), _ro_src_objs(), _src_obj_table(INITIAL_TABLE_SIZE) {
-  assert(_current == NULL, "must be");
-  _current = this;
-
+ArchiveBuilder::ArchiveBuilder() :
+  _current_dump_space(NULL),
+  _buffer_bottom(NULL),
+  _last_verified_top(NULL),
+  _num_dump_regions_used(0),
+  _other_region_used_bytes(0),
+  _requested_static_archive_bottom(NULL),
+  _requested_static_archive_top(NULL),
+  _requested_dynamic_archive_bottom(NULL),
+  _requested_dynamic_archive_top(NULL),
+  _mapped_static_archive_bottom(NULL),
+  _mapped_static_archive_top(NULL),
+  _buffer_to_requested_delta(0),
+  _mc_region("mc", MAX_SHARED_DELTA),
+  _rw_region("rw", MAX_SHARED_DELTA),
+  _ro_region("ro", MAX_SHARED_DELTA),
+  _rw_src_objs(),
+  _ro_src_objs(),
+  _src_obj_table(INITIAL_TABLE_SIZE),
+  _num_instance_klasses(0),
+  _num_obj_array_klasses(0),
+  _num_type_array_klasses(0),
+  _total_closed_heap_region_size(0),
+  _total_open_heap_region_size(0),
+  _estimated_metaspaceobj_bytes(0),
+  _estimated_hashtable_bytes(0),
+  _estimated_trampoline_bytes(0)
+{
   _klasses = new (ResourceObj::C_HEAP, mtClassShared) GrowableArray<Klass*>(4 * K, mtClassShared);
   _symbols = new (ResourceObj::C_HEAP, mtClassShared) GrowableArray<Symbol*>(256 * K, mtClassShared);
   _special_refs = new (ResourceObj::C_HEAP, mtClassShared) GrowableArray<SpecialRefInfo>(24 * K, mtClassShared);
 
-  _num_instance_klasses = 0;
-  _num_obj_array_klasses = 0;
-  _num_type_array_klasses = 0;
-  _alloc_stats = new (ResourceObj::C_HEAP, mtClassShared) DumpAllocStats;
-
-  _mc_region = mc_region;
-  _rw_region = rw_region;
-  _ro_region = ro_region;
-
-  _num_dump_regions_used = 0;
-
-  _estimated_metaspaceobj_bytes = 0;
-  _estimated_hashtable_bytes = 0;
-  _estimated_trampoline_bytes = 0;
-
-  _requested_static_archive_bottom = NULL;
-  _requested_static_archive_top = NULL;
-  _mapped_static_archive_bottom = NULL;
-  _mapped_static_archive_top = NULL;
-  _requested_dynamic_archive_bottom = NULL;
-  _requested_dynamic_archive_top = NULL;
-  _buffer_to_requested_delta = 0;
+  assert(_current == NULL, "must be");
+  _current = this;
```

Comment:
These lines used to be at the beginning of the function. Any reasons why they are moved?

Reply:
I think it's better to set `_current` after the instance has been fully initialized.